torch/optim/lars.py

```diff
@@ -0,0 +1,214 @@
+from .optimizer import Optimizer, _use_grad_for_differentiable, _differentiable_doc, _maximize_doc  # type: ignore[attr-defined]
+import torch
+from typing import List, Optional
+from torch import Tensor
+
+__all__ = ["LARS", "lars"]
+
+
+class LARS(Optimizer):
+    def __init__(
+        self,
+        params,
+        lr=1e-3,
+        momentum: float = 0,
+        dampening: float = 0,
+        weight_decay: float = 0,
+        nesterov: bool = False,
+        *,
+        trust_coefficient: float = 0.001,
+        eps: float = 1e-8,
+        maximize: bool = False,
+        differentiable: bool = False,
+    ):
+        if lr < 0.0:
+            raise ValueError(f"Invalid learning rate: {lr}")
+        if weight_decay < 0.0:
+            raise ValueError(f"Invalid weight decay value: {weight_decay}")
+        if momentum < 0.0:
+            raise ValueError(f"Invalid momentum value: {momentum}")
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+
+        defaults = dict(
+            lr=lr,
+            momentum=momentum,
+            dampening=dampening,
+            weight_decay=weight_decay,
+            nesterov=nesterov,
+            trust_coefficient=trust_coefficient,
+            eps=eps,
+            maximize=maximize,
+            differentiable=differentiable,
+        )
+
+        super().__init__(params, defaults)
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault("nesterov", False)
+            group.setdefault("maximize", False)
+            group.setdefault("differentiable", False)
+
+    def _init_group(self, group, params_with_grad, grads, momentum_buffer_list):
```

Comment:
I notice that has_sparse_grad is excluded from this implementation so far. What are your plans with the variable?

Wanted to give you the heads up that we are planning on deprecating has_sparse_grad, because our main use of it is to determine whether we could use the foreach implementation or not for it. Instead of barring all params from going through the foreach implementation because of one sparse grad, the new ideal is to group the tensors by device and to put everything that cannot be foreach'd into the cpu bucket, so that we can maximally enjoy the speed of foreach.

Reply:
To be honest, I skipped it. My idea was to complete at least the basic implementation, then implement more "advanced" parameters, like `sparse` and `foreach`.

Reply:
yes, I wanted to give you more context regarding the deprecation of has_sparse_grad (as we plan to remove it for all optims as well)